### PR TITLE
Initial commit for dwarf

### PR DESCRIPTION
**Pull request by P4N74**
NOT FOR MERGE:
There are currently 2 methods implemented (C-like output and pf output) in the code. Need to pick one and build on top of it.

Also I will remove strng struct and instead use r_str_buf before final commit, but currently I am not focusing on fixing that and instead working on the main part. Also code has real bad error handling which needs to be fixed (with return value and default value in case something fails).

For struct :
```
struct abc {
	int a;
	char b;
	int c;
	char *d;
	int e; 
	int f;
};
```

pf output:
```
pf.abc ic...i:pii a b c d e f
```

and C output:
```
struct abc {
	int a;
	char b;
	ut16 temp0[1];
	ut8 temp1[1];
	int c;
	ut32 temp2[1];
	char *d;
	int e;
	int f;
};
```

### CODE REVIEW

File: dwarf/main.c [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+		}
+	} else {
+		strcpy (dst->buffer, src1.buffer);
+		if (strcmp (tmp_buf, "")) {
+			if (strcmp (dst->buffer, ""))
+				strcat (dst->buffer, delim);
+			strcat (dst->buffer, src2.buffer);
+		}
+	}
+	if (tmp_buf)
```
Comment (radare) [inline dwarf/main.c]: Remove that if. Free already checks for null

File: dwarf/main.c [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+		free (tmp_buf);
+}
+
+strng* strng_new (char *buf) {
```
Comment (radare) [inline dwarf/main.c]: All those functions must be static

File: dwarf/main.c [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+	strng *result = malloc(sizeof(strng));
+	result->bufsize = 2*strlen(buf);
+	result->buffer = malloc (2*strlen(buf));
```
Comment (radare) [inline dwarf/main.c]: Spaces before every (

Comment (radare) [inline dwarf/main.c]: Dont compute strlen*2 twice. Reuse the bufsize

File: dwarf/main.c [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+	fd = open(argv[1], O_RDONLY);
+	if (fd < 0)
+		return -1;
+	res = dwarf_init (fd, DW_DLC_READ, errhand, errarg, &dbg, &error);
+	if (res != DW_DLV_OK)
+		return -1;
+	res = read_cu_list(dbg);
+	if (res != 0)
+		return -1;
+	dwarf_dealloc(dbg, error, DW_DLA_ERROR);
```
Comment (radare) [inline dwarf/main.c]: Missing space before (

File: dwarf/main.c [hunk 5]
```diff
@@ -0,0 +1,1011 @@
...
+		if (res == DW_DLV_ERROR) {
+			continue;
+		}
+		if (res == DW_DLV_NO_ENTRY) {
+			return 0;
+		}
+
+		res = dwarf_siblingof (dbg, NULL, &cu_die, &error);
+		if (res == DW_DLV_ERROR) {
+			printf ("Error in dwarf_siblingof on CU die\n");
```
Comment (radare) [inline dwarf/main.c]: Wrong indent (spaces vs tabs) use only tabs

File: dwarf/main.c [hunk 6]
```diff
@@ -0,0 +1,1011 @@
...
+		    return 1;
+		}
+		if (res == DW_DLV_NO_ENTRY) {
+			/* Impossible case. */
+			printf ("no entry! in dwarf_siblingof on CU die.\n");
+		    return -1;
+		}
+		get_die_and_siblings (dbg, cu_die, 0);//error handling??
+	    dwarf_dealloc (dbg, cu_die, DW_DLA_DIE);
```
Comment (radare) [inline dwarf/main.c]: Wring indent

File: dwarf/main.c [hunk 7]
```diff
@@ -0,0 +1,1011 @@
...
+		if (res == DW_DLV_NO_ENTRY) {
+			break;
+		}
+
+		if (cur_die != in_die)
+			dwarf_dealloc (dbg, cur_die, DW_DLA_DIE);
+		cur_die = sib_die;
+		print_die_data (dbg, cur_die, in_level);
+	}
+	if (cur_die != in_die)
```
Comment (radare) [inline dwarf/main.c]: Add braces around oneliner ifs

File: dwarf/main.c [hunk 8]
```diff
@@ -0,0 +1,1011 @@
...
+	return -1;
+}
+
+static int get_array_length(Dwarf_Debug dbg, Dwarf_Die die, unsigned int *len) {
+	Dwarf_Attribute attr = 0;
+	Dwarf_Error error = 0;
+	Dwarf_Half tag = 0;
+	Dwarf_Unsigned bound = 0;
+	int res = DW_DLV_ERROR;
+
```
Comment (radare) [inline dwarf/main.c]: Unused assignment

File: dwarf/main.c [hunk 9]
```diff
@@ -0,0 +1,1011 @@
...
+			return -1;
+		}
+		if (res == DW_DLV_NO_ENTRY) {
+			dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+			dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+			*len = 0;
+			return -1;
+		}
+		res = get_number (attr, &bound);
+	    if (res == 0){
```
Comment (radare) [inline dwarf/main.c]: Wrong indent and missing space before {

File: dwarf/main.c [hunk 10]
```diff
@@ -0,0 +1,1011 @@
...
+		if (res == DW_DLV_NO_ENTRY) {
+			typename = "<no DW_AT_name attr>";
+			localtype = 1;
+		}
+		strng *tmpstring = strng_new (typename);
+		strngcat(p_name, *tmpstring, *p_name, " ");
+		strng_free (tmpstring);
+		break;
+	}
+	case DW_TAG_typedef: {
```
Comment (radare) [inline dwarf/main.c]: No need for those braces

File: dwarf/main.c [hunk 11]
```diff
@@ -0,0 +1,1011 @@
...
+		return -1;//exit(1);
+	}
+
+	res = dwarf_tag (typedie, &tag, &error);
+	if (res != DW_DLV_OK) {
+		return -1;//exit(1);
+	}
+
+	switch (tag) {
+	case DW_TAG_base_type: {
```
Comment (radare) [inline dwarf/main.c]: No need for those braces

File: dwarf/main.c [hunk 12]
```diff
@@ -0,0 +1,1011 @@
...
+	dwarf_dealloc (dbg, typedie, DW_DLA_DIE);
+	dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+	return 0;
+}
+
+int get_type_size (Dwarf_Debug dbg, Dwarf_Die die, unsigned int *size) {
+	Dwarf_Attribute attr = 0;
+	Dwarf_Off offset;
+	Dwarf_Die typedie = 0;
+    Dwarf_Error error = 0;
```
Comment (radare) [inline dwarf/main.c]: Wrong indent

File: dwarf/main.c [hunk 13]
```diff
@@ -0,0 +1,1011 @@
...
+	Dwarf_Unsigned sz = 0;
+	int res = DW_DLV_ERROR;
+	Dwarf_Half tag = 0;
+
+	res = dwarf_attr (die, DW_AT_type, &attr, &error);
+	if (res == DW_DLV_ERROR) {
+		dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+		exit(1);
```
Comment (radare) [inline dwarf/main.c]: Dont exit in library code. Only main() can do that. Control the error in higher layers

File: dwarf/main.c [hunk 14]
```diff
@@ -0,0 +1,1011 @@
...
+	res = dwarf_global_formref (attr, &offset, &error);
+	if (res == DW_DLV_ERROR) {
+		*size = 0;
+		exit(1);
+	}
+
+	res = dwarf_offdie (dbg, offset, &typedie, &error);
+	if (res != DW_DLV_OK) {
+		dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+		dwarf_dealloc (dbg, error, DW_DLA_ERROR);
```
Comment (radare) [inline dwarf/main.c]: Same here

File: dwarf/main.c (hunk 13 above)
Comment (P4N74) [inline dwarf/main.c]: yes, I will remove all the exit. Currently using them so that I can be sure that code runs on the entire file.

**Comment by P4N74** [comment]
I will fix all the comment. But main part is to fix what kind of output is required as of now

**Comment by radare** [comment]
Yep i will review this once merged, after the code review

File: dwarf/main.c (hunk 5 above)
Comment (P4N74) [inline dwarf/main.c]: sorry, but I don't see indentation error from line 119 to 141.

File: dwarf/main.c (hunk 8 above)
Comment (P4N74) [inline dwarf/main.c]: it is just to be on safe side to have error if the return value is not set. I can change it if it actually doesn't matter.

**Comment by radare** [comment]
Please update this pr

**Pull request merged**